File: app/src/main/java/dev/yiray/qickmvivm/ui/home/HomeFragment.java
package dev.yiray.qickmvivm.ui.home;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;

import com.jakewharton.rxbinding4.view.RxView;
import com.jakewharton.rxbinding4.widget.RxCompoundButton;
import com.jakewharton.rxbinding4.widget.RxTextView;

import java.util.concurrent.TimeUnit;

import dev.yiray.qickmvivm.base.BaseMVIVMFragment;
import dev.yiray.qickmvivm.databinding.FragmentHomeBinding;
import dev.yiray.qickmvivm.ui.ViewModelFactory;
import io.reactivex.rxjava3.core.Observable;

public class HomeFragment extends BaseMVIVMFragment<HomeView.Action, HomeViewState, HomeViewModel>
    implements HomeView, HomeView.Action{

    private FragmentHomeBinding binding;
    private Coordinator coordinator;

    public interface Coordinator {
        void toTaskList();
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        try {
            coordinator = (Coordinator) context;
        } catch (ClassCastException e) {
            throw new ClassCastException("MUST implement Coordination: " + context.toString());
        }
    }

    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {

        binding = FragmentHomeBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        mViewModel.bindIntent(this);
    }

    @Override
    protected HomeViewModel bindViewModel() {
        // Bind Activity's Lifecycle or you can use
        // navController.getViewModelStoreOwner(*your_nested_graph_id) to bind with nested graph
        return new ViewModelProvider(requireActivity(), new ViewModelFactory()).get(HomeViewModel.class);
    }

    @Override
    public void displayErrorMsg(String message) {
        Toast.makeText(getContext(), message, Toast.LENGTH_LONG).show();
    }

    @Override
    public void render(HomeViewState state) {
        binding.textHome.setText(state.toViewState().getText());
        binding.buttonNext.setEnabled(state.toViewState().getChecked());
    }

    @Override
    public void handleSideEffect(HomeViewState state, HomeViewState sideEffect) {
        if (sideEffect instanceof HomeViewState.NextPage) {
            coordinator.toTaskList();
        }
    }

    @Override
    public Observable<CharSequence> observableInput() {
        return RxTextView.textChanges(binding.input);
    }

    @Override
    public Observable<Boolean> observableCheckBox() {
        return RxCompoundButton.checkedChanges(binding.checkboxPrepared);
    }

    @Override
    public Observable<Boolean> observableNextPage() {
        return RxView.clicks(binding.buttonNext)
                .debounce(200, TimeUnit.MILLISECONDS)
                .map(e -> true);
    }

    @Override
    public Observable<UpdateTask> observableUpdated() {
        return Observable.empty();
    }
}

File: app/src/main/java/dev/yiray/livedatamvi/ui/home/HomeFragment.kt
package dev.yiray.livedatamvi.ui.home

import android.content.Context
import dev.yiray.livedatamvi.ui.home.HomeView.UpdateTask
import dev.yiray.livedatamvi.base.BaseMVIVMFragment
import android.view.LayoutInflater
import android.view.ViewGroup
import android.os.Bundle
import android.text.Editable
import androidx.lifecycle.ViewModelProvider
import dev.yiray.livedatamvi.ui.ViewModelFactory
import android.widget.Toast
import dev.yiray.livedatamvi.ui.home.HomeViewState.NextPage
import android.view.View
import androidx.core.widget.doOnTextChanged
import dev.yiray.livedatamvi.databinding.FragmentHomeBinding
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.flow
import java.lang.ClassCastException

class HomeFragment : BaseMVIVMFragment<HomeView.Action, HomeViewState, HomeViewModel>(), HomeView,
    HomeView.Action {
    private var binding: FragmentHomeBinding? = null
    private var coordinator: Coordinator? = null

    interface Coordinator {
        fun toTaskList()
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        coordinator = try {
            context as Coordinator
        } catch (e: ClassCastException) {
            throw ClassCastException("MUST implement Coordination: $context")
        }
    }

    override fun onStart() {
        super.onStart()
        mViewModel!!.bindIntent(this)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?, savedInstanceState: Bundle?
    ): View {
        binding = FragmentHomeBinding.inflate(inflater, container, false)
        return binding!!.root
    }

    override fun bindViewModel(): HomeViewModel {
        // Bind Activity's Lifecycle or you can use
        // navController.getViewModelStoreOwner(*your_nested_graph_id) to bind with nested graph
        return ViewModelProvider(
            requireActivity(),
            ViewModelFactory()
        ).get(HomeViewModel::class.java)
    }

    override fun displayErrorMsg(message: String?) {
        Toast.makeText(context, message, Toast.LENGTH_LONG).show()
    }

    override fun render(state: HomeViewState) {
        val text = state.toViewState()?.text
        val isChecked = state.toViewState()!!.checked
        binding!!.textHome.text = text

        if (text != null && binding!!.input.text.toString() != text) {
            binding!!.input.text = Editable.Factory.getInstance().newEditable(text)
        }

        binding!!.checkboxPrepared.isChecked = isChecked
        binding!!.buttonNext.isEnabled = isChecked
    }

    override fun handleSideEffect(state: HomeViewState, sideEffect: HomeViewState) {
        if (sideEffect is NextPage) {
            coordinator!!.toTaskList()
        }
    }

    override fun observableInput(): Flow<CharSequence> {
        return callbackFlow {
            binding!!.input.doOnTextChanged { text, _, _, _ ->
                if (text != null) {
                    trySend(text)
                }
            }

            awaitClose { binding!!.input.addTextChangedListener(null) }
        }
    }

    override fun observableCheckBox(): Flow<Boolean> {
        return callbackFlow {
            binding!!.checkboxPrepared.setOnCheckedChangeListener { _, isChecked ->
                trySend(isChecked)
            }

            awaitClose { binding!!.checkboxPrepared.setOnCheckedChangeListener(null) }
        }
    }

    override fun observableNextPage(): Flow<Boolean> {
        return callbackFlow {
            binding!!.buttonNext.setOnClickListener {
                trySend(true)
            }
            awaitClose { binding!!.buttonNext.setOnClickListener(null) }
        }
    }

    override fun observableUpdated(): Flow<UpdateTask> {
        return flow { }
    }
}